src/QsCompiler/SyntaxProcessor/VerificationTools.fs

```diff
@@ -108,4 +108,14 @@ type ResolvedType with
         | ArrayType bt -> Some bt
         | _ -> None
 
-
+/// Converts the range relative to the QsLocation's offset into a range relative to the root node.
+let internal rangeRelativeToRoot (location : QsLocation) =
```

Comment:
It's the responsibility of the SyntaxProcessor to generate diagnostics. Because diagnostics require a range of type `QsPositionInfo * QsPositionInfo`, it means that the SyntaxProcessor must convert `QsLocation` into `QsPositionInfo * QsPositionInfo`. We can delay the computation by returning `(int * int) * QsCompilerDiagnostic` pairs, but this is an extra layer of indirection that is not necessary, and overuse of tuples (especially nested tuples) can decrease code clarity.

Reply:
If we want to delegate responsibility, we can say that `QsLocation` (defined in DataStructures) should know how to combine its `Offset` and `Range` - it has all the information needed to do that.

Reply:
It would need to know though which one is one based, and which one is zero based. I much more prefer QsLocation actually taking on that logic rather than the syntax processor, but it still means that now different pieces of code rely on knowledge about the implementation in another project. There are other ways to deal with code clarity, like returning a QsLocation instead of a tuple for example.

Reply:
It would not rely on implementation details if `QsLocation` specified whether the offset and/or range was zero- or one-based. Then the project that creates `QsLocation` would convert it to the correct representation and any other projects that consume `QsLocation` can rely on the fact that it's zero-based (or one-based).

Carrying unnormalized data throughout the compiler makes things more complicated than they need to be, IMO. For example, equality is not well-defined. A project could choose to create one-based `QsLocation`s and another project could choose to create zero-based `QsLocation`s. Because the representation is undefined, these two `QsLocation`s are not equal even though conceptually they may represent the same location.

If `QsLocation` required a particular representation (and it could provide convenient constructors for this - for example `QsLocation.FromOneBased` or something like that), then equality and combining range and offset would both be well-defined operations that did not rely on implementation details of other projects.

Reply:
As an example, consider the [`Complex`](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.complex?view=netcore-3.1) type. Technically, it's just a pair of doubles, and you could say that the representation is undefined (either Cartesian or polar). But the `Complex` constructor requires Cartesian coordinates - that's documented and part of the API. It also provides a convenient alternative constructor `Complex.FromPolarCoordinates` if you want to use polar coordinates instead.

As a result, manipulating complex numbers is well-defined regardless of whether they were created using Cartesian or polar coordinates, and this allows `Complex` to do the conversions internally to provide properties that expose the complex number as either Cartesian (`Real`, `Imaginary`) or polar (`Magnitude`, `Phase`).

Reply:
Based on our conversation, I created #506 to track refactoring position/ranges. For now, I'll move this to DiagnosticTools.

Reply:
Perfect, thanks!